server.go

```diff
@@ -34,3 +36,22 @@ func (s *Server) SetStorage(stg *Storage) error {
 	s.Stats.TotalURL = uint64(len(stg.URLPairs))
 	return nil
 }
+
+func (s *Server) Storage() *Storage {
+	return s.Kvs.Dump()
+}
+
+func (s *Server) ShutdownHandler(args *cmdArgs) {
+	termChan := make(chan os.Signal)
+	signal.Notify(termChan, os.Interrupt, os.Kill)
+	go func() {
+		<-termChan // Blocks here until interrupted
+		err := s.Storage().SaveStorageFile(args.file)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Cannot save storage: %v\n", err)
+			os.Exit(1)
+		}
+		fmt.Println("Save storage")
+		os.Exit(0)
+	}()
```

Comment:
I'm not sure the fact that the Server directly interacts with the OS (to catch the signal) is the best way here. If you think about it, we may want to add in the future, for example, to automatically save the current storage to disk every 5s, and not only at shutdown. Also if you were to add a test for `Server.ShutdownHandler` you would need to send an OS signal within the test. That's not ideal.

One proposition would be to implement the signal catching goroutine outside of the Server, and just have that goroutine calls `Server.Shutdown` (or whatever name you chose), thus having the same result but with decoupled responsibility.

Advice:
Always try to organize your code as if tomorrow you'd have to transform it to a library.  In a sense, apart from importing this hypothetical url-shortner server library from your `main` the code would be very similar.

Thinking in this way helps you make decisions about what to make public/private (in Go world: export/unexport), how to design your public API. Also, a library should never takes the responsibility of calling `os.Exit()` and thus terminating the process, if there's an error, return it, let it bubble up the call stack until the `main` package/function decides whether or not it can recover from it or should die.

In our specific case, if the signal handling goroutines was out the server, ShutdownHandler could simply return an error in case something happened while saving the file.